google_guest_agent/instance_setup.go

```diff
@@ -195,6 +219,341 @@ func agentInit(ctx context.Context) {
+
+	for cpu := queueNum; cpu < totalCPUs; cpu += numQueues {
+		dword := cpu / 32
+		xps[dword] |= 1 << uint32(cpu-(32*dword))
+	}
+	// Linux xps_cpus requires a hex number with commas every 32 bits. It ignores
+	// all bits above # cpus, so write a list of comma separated 32 bit hex values
+	// with a comma between dwords.
+	var xpsDwords []string
+	for i := 0; i < numwords; i++ {
+		xpsDwords = append([]string{fmt.Sprintf("%08x", xps[i])}, xpsDwords...)
+	}
+	return strings.Join(xpsDwords, ",")
+}
+
+func copyFile(src string, dest string) error {
+	input, err := ioutil.ReadFile(src)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(dest, input, 0644)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func setQueueNumForDevice(dev string) error {
+	dev = path.Base(dev)
+	irqDirs, err := filepath.Glob(irqDirPath)
+	if err != nil {
+		return err
+	}
+	for _, irq := range irqDirs {
+		smpAffinity := irq + "/smp_affinity_list"
+		if !isFile(smpAffinity) {
+			continue
+		}
+		virtionetIntxDir := irq + "/" + dev
+		virtionetMsixDirRegex := ".*/" + dev + "-(input|output)\\.([0-9]+)$"
+		if isDir(virtionetIntxDir) {
+			// All virtionet intx IRQs are delivered to CPU 0
+			logger.Infof("Setting %s to 01 for device %s.", smpAffinity, dev)
+			if err := ioutil.WriteFile(smpAffinity, []byte("01"), 0644); err != nil {
+				logger.Warningf("Could not write 01 to %s for %s.", smpAffinity, dev)
+			}
+			continue
+		}
+		// Not virtionet intx, probe for MSI-X
+		var virtionetMsixFound bool
+		irqDevs, err := filepath.Glob(irq + "/" + dev + "*")
+		if err != nil {
+			return err
+		}
+		var queueNum int
+		for _, entry := range irqDevs {
+			r, _ := regexp.Compile(virtionetMsixDirRegex)
+			if r.MatchString(entry) {
+				virtionetMsixFound = true
+				queueNum, err = strconv.Atoi(r.FindStringSubmatch(entry)[2])
+				if err != nil {
+					return err
+				}
+			}
+		}
+		affinityHint := irq + "/affinity_hint"
+		if !virtionetMsixFound || !isFile(affinityHint) {
+			continue
+		}
+
+		//Set the IRQ CPU affinity to the virtionet-initialized affinity hint
+		logger.Infof("Setting %s to %d for device %s.", smpAffinity, queueNum, dev)
+		if err = ioutil.WriteFile(smpAffinity, []byte(strconv.Itoa(queueNum)), 0644); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func configNVME(totalCPUs int) error {
+	var currentCPU = 0
+	devices, err := filepath.Glob(nvmeDevice)
+	if err != nil {
+		return err
+	}
+	for _, dev := range devices {
+		if !isDir(dev) {
+			continue
+		}
+		irqs, err := filepath.Glob(dev + "/msi_irqs/*")
+		if err != nil {
+			return err
+		}
+
+		for _, irqInfo := range irqs {
+			if !isFile(irqInfo) {
+				continue
+			}
+			currentCPU := currentCPU % totalCPUs
+			cpuMask := 1 << currentCPU
+			irq := path.Base(irqInfo)
+			logger.Infof("Setting IRQ %s smp_affinity to %d", irq, cpuMask)
+			if err := ioutil.WriteFile("/proc/irq/"+irq+"/smp_affinity", []byte(strconv.Itoa(cpuMask)), 0644); err != nil {
```

Comment:
maybe a nitpick but you use + here and Sprintf below in configSCSI, maybe stick to one

Reply:
should stick to Sprintf.